domain/externalcontroller/state/state.go

```diff
@@ -0,0 +1,94 @@
+// Copyright 2023 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package state
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+
+	"github.com/juju/collections/transform"
+	"github.com/juju/errors"
+	"github.com/juju/utils/v3"
+
+	"github.com/juju/juju/core/crossmodel"
+	"github.com/juju/juju/database"
+	"github.com/juju/juju/domain"
+)
+
+type State struct {
+	*domain.StateBase
+}
+
+func NewState(factory domain.DBFactory) *State {
+	return &State{
+		StateBase: domain.NewStateBase(factory),
+	}
+}
+
+func (st *State) UpdateExternalController(ctx context.Context, ci crossmodel.ControllerInfo, modelUUIDs []string) error {
+	ec := ExternalController{
+		ID:     ci.ControllerTag.Id(),
+		Alias:  ci.Alias,
+		Addrs:  ci.Addrs,
+		CACert: ci.CACert,
+		Models: modelUUIDs,
+	}
+
+	db, err := st.DB()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	err = db.Txn(ctx, func(ctx context.Context, tx *sql.Tx) error {
+		q := `
+INSERT INTO external_controller (uuid, alias, ca_cert)
+VALUES (?, ?, ?)
+  ON CONFLICT(uuid) DO UPDATE SET alias=?, ca_cert=?`[1:]
+
+		if _, err := tx.ExecContext(ctx, q, ec.ID, ec.Alias, ec.CACert, ec.Alias, ec.CACert); err != nil {
+			return errors.Trace(err)
+		}
+
+		q = fmt.Sprintf(`
+DELETE FROM external_controller_address
+WHERE  controller_uuid = ?
+AND    address NOT IN (%s)`[1:], database.SliceToPlaceholder(ec.Addrs))
```

Comment:
I was curious what happens if Addrs is empty, but it seems to do what you expect:
```
sqlite> CREATE TABLE foo (id string primary key); CREATE TABLE bar (id string, val int, foreign key (id) references foo(id));
sqlite> insert into foo values ("a");
sqlite> insert into foo values ("b");
sqlite> insert into bar values ("a", 1);
sqlite> insert into bar values ("a", 2);
sqlite> select * from bar;
a|1
a|2
sqlite> delete from bar where id="a" and val not in ();
sqlite> select * from bar;
sqlite>
```

Reply:
Yep. `TestUpdateExternalControllerUpdateModel` flexes this case.